File: py/abi.py
from algosdk.v2client.algod import AlgodClient
from algosdk.atomic_transaction_composer import (
    AccountTransactionSigner,
    AtomicTransactionComposer,
    TransactionWithSigner,
)
from algosdk.transaction import PaymentTxn, AssetCreateTxn
from algosdk.abi import Contract
from algosdk.logic import get_application_address

from sandbox import get_accounts

client = AlgodClient("a" * 64, "http://localhost:4001")


addr, sk = get_accounts()[0]

with open("../contract.json") as f:
    js = f.read()

with open("../.app_id") as f:
    app_id = int(f.read())

app_addr = get_application_address(app_id)

c = Contract.from_json(js)

signer = AccountTransactionSigner(sk)
sp = client.suggested_params()


box_comp = AtomicTransactionComposer()
box_name = b"cool_box"
box_comp.add_transaction(
    TransactionWithSigner(PaymentTxn(addr, sp, app_addr, 1_000_000_000), signer=signer),
)
box_comp.add_method_call(
    app_id,
    c.get_method_by_name("box_write"),
    addr,
    sp,
    signer,
    method_args=[box_name, (123, 456)],
    boxes=[(0, box_name)],
)
box_comp.add_method_call(
    app_id,
    c.get_method_by_name("box_read"),
    addr,
    sp,
    signer,
    method_args=[box_name],
    # technically not needed since we already ref it in
    # the previous txn in the same group
    boxes=[(0, box_name)],
)
result = box_comp.execute(client, 4)
print(f"box_read returned: {result.abi_results[-1].return_value}")


comp = AtomicTransactionComposer()

comp.add_method_call(
    app_id, c.get_method_by_name("add"), addr, sp, signer, method_args=[1, 1]
)
comp.add_method_call(
    app_id, c.get_method_by_name("sub"), addr, sp, signer, method_args=[3, 1]
)
comp.add_method_call(
    app_id, c.get_method_by_name("div"), addr, sp, signer, method_args=[4, 2]
)
comp.add_method_call(
    app_id, c.get_method_by_name("mul"), addr, sp, signer, method_args=[3, 2]
)
comp.add_method_call(
    app_id, c.get_method_by_name("qrem"), addr, sp, signer, method_args=[27, 5]
)

comp.add_method_call(
    app_id,
    c.get_method_by_name("reverse"),
    addr,
    sp,
    signer,
    method_args=["desrever yllufsseccus"],
)

ptxn = TransactionWithSigner(PaymentTxn(addr, sp, addr, 10000), signer)
comp.add_method_call(
    app_id,
    c.get_method_by_name("txntest"),
    addr,
    sp,
    signer,
    method_args=[10000, ptxn, 1000],
)

comp.add_method_call(
    app_id, c.get_method_by_name("manyargs"), addr, sp, signer, method_args=[2] * 20
)


comp.add_method_call(
    app_id,
    c.get_method_by_name("min_bal"),
    addr,
    sp,
    signer,
    method_args=["SKCBRBKPIGY5LI2OU63IE5LMNQ5BVVOKPHWTPPWFQOI4NG4TI35SLAA3JQ"],
)

comp.add_method_call(
    app_id,
    c.get_method_by_name("concat_strings"),
    addr,
    sp,
    signer,
    method_args=[["this", "string", "is", "joined"]],
)

# Useable with abi_types.py contract
# comp.add_method_call(
#     app_id,
#     c.get_method_by_name("concat_dynamic_arrays"),
#     addr,
#     sp,
#     signer,
#     method_args=[[1, 2, 3], [4, 5, 6]],
# )
#
# comp.add_method_call(
#     app_id,
#     c.get_method_by_name("concat_static_arrays"),
#     addr,
#     sp,
#     signer,
#     method_args=[[1, 2, 3], [4, 5, 6]],
# )
#
#
# comp.add_method_call(
#     app_id,
#     c.get_method_by_name("concat_dynamic_string_arrays"),
#     addr,
#     sp,
#     signer,
#     method_args=[["a", "b", "c"], ["d", "e", "f"]],
# )


# drr = comp.dryrun(client)
# for txn in drr.trace.txns:
#    if txn.app_call_rejected():
#        print(txn.app_trace())

resp = comp.execute(client, 2)
for result in resp.abi_results:
    # print(result.decode_error)
    # print(result.raw_value.hex())
    print(f"{result.method.name} => {result.return_value}")

File: js/abi.ts
import algosdk, {
  decodeAddress,
  makePaymentTxnWithSuggestedParamsFromObject,
  Transaction,
} from "algosdk";
import * as fs from "fs";
import { Buffer } from "buffer";
import { getAccounts } from "./sandbox";

const algod_token =
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
const algod_host = "http://127.0.0.1";
const algod_port = "4001";

(async function () {
  // Create a client to communicate with local node
  const client = new algosdk.Algodv2(algod_token, algod_host, algod_port);

  // Get account from sandbox
  const accounts = await getAccounts();
  const acct = accounts[0];

  // Read in the local contract.json file
  const buff = fs.readFileSync("../contract.json");

  // Parse the json file into an object, pass it to create an ABIContract object
  const contract = new algosdk.ABIContract(JSON.parse(buff.toString()));

  // Get the app id for the app we created with `./manage.sh` earlier
  const appId = parseInt(fs.readFileSync("../.app_id").toString());

  // Get sp once, we'll reuse it
  const sp = await client.getTransactionParams().do();

  // Fund the app address so it can create boxes
  const ptxn = makePaymentTxnWithSuggestedParamsFromObject({
    from: acct.addr,
    suggestedParams: sp,
    to: algosdk.getApplicationAddress(appId),
    amount: 1_000_000_000,
  }).signTxn(acct.sk);
  await client.sendRawTransaction(ptxn).do();

  // We initialize the common parameters here, they'll be passed to all the transactions
  // since they happen to be the same
  const commonParams = {
    appID: appId,
    sender: acct.addr,
    suggestedParams: sp,
    signer: algosdk.makeBasicAccountTransactionSigner(acct),
  };

  // Write then read a box
  const boxComp = new algosdk.AtomicTransactionComposer();
  const boxName = new Uint8Array(Buffer.from("cool_box"));
  boxComp.addMethodCall({
    method: contract.getMethodByName("box_write"),
    methodArgs: [boxName, [123, 456]],
    boxes: [{ appIndex: 0, name: boxName }],
    ...commonParams,
  });
  boxComp.addMethodCall({
    method: contract.getMethodByName("box_read"),
    methodArgs: [boxName],
    boxes: [{ appIndex: 0, name: boxName }],
    ...commonParams,
  });
  const boxResult = await boxComp.execute(client, 4);
  console.log("Box contents: ", boxResult.methodResults[1].returnValue);

  // Simple ABI Calls with standard arguments, return type
  const comp = new algosdk.AtomicTransactionComposer();
  comp.addMethodCall({
    method: contract.getMethodByName("add"),
    methodArgs: [1, 1],
    ...commonParams,
  });
  comp.addMethodCall({
    method: contract.getMethodByName("sub"),
    methodArgs: [3, 1],
    ...commonParams,
  });
  comp.addMethodCall({
    method: contract.getMethodByName("div"),
    methodArgs: [4, 2],
    ...commonParams,
  });
  comp.addMethodCall({
    method: contract.getMethodByName("mul"),
    methodArgs: [3, 3],
    ...commonParams,
  });

  //Tuple return type
  comp.addMethodCall({
    method: contract.getMethodByName("qrem"),
    methodArgs: [27, 5],
    ...commonParams,
  });

  // String return type
  comp.addMethodCall({
    method: contract.getMethodByName("reverse"),
    methodArgs: [Buffer.from("desrever yllufsseccus")],
    ...commonParams,
  });

  // Transaction being passed as an argument, this removes the transaction from the
  // args list, but includes it in the atomic grouped transaction
  comp.addMethodCall({
    method: contract.getMethodByName("txntest"),
    methodArgs: [
      10000,
      {
        txn: new Transaction({
          from: acct.addr,
          to: acct.addr,
          amount: 10000,
          ...sp,
        }),
        signer: algosdk.makeBasicAccountTransactionSigner(acct),
      },
      1000,
    ],
    ...commonParams,
  });

  // Here we call with 20 arguments to demonstrate Tuple encoding of any arguments past index 14
  comp.addMethodCall({
    method: contract.getMethodByName("manyargs"),
    methodArgs: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
    ...commonParams,
  });

  // Pass in an account by address, the atc will take care of mapping this correctly
  comp.addMethodCall({
    method: contract.getMethodByName("min_bal"),
    methodArgs: ["FHWVNNZOALOSBKYFKEUIZC56SGPLLAREZFFWLXCPBBVVISXDLPTRFR7EIQ"],
    ...commonParams,
  });

  // Dynamic argument types are supported (undefined length array)
  comp.addMethodCall({
    method: contract.getMethodByName("concat_strings"),
    methodArgs: [["this", "string", "is", "joined"]],
    ...commonParams,
  });

  // This is not necessary to call but it is helpful for debugging
  // to see what is being sent to the network
  //for(const gtxn of comp.buildGroup()){ console.log(gtxn.txn.appArgs) }

  // We can also dryrun the group
  // const res = await comp.dryrun(client)
  // for(const tx of res.methodResults){ console.log(tx.returnValue) }

  // Finally, execute the composed group and print out the results
  const results = await comp.execute(client, 2);
  for (const result of results.methodResults) {
    console.log(`${result.method.name} => ${result.returnValue}`);
  }
})();
